daemon/emer-circular-file.c

```diff
@@ -0,0 +1,825 @@
+                   priv->size);
+      return FALSE;
+    }
+
+  goffset little_endian_key =
+    g_key_file_get_int64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                          HEAD_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  priv->head = swap_bytes_64_if_big_endian (little_endian_key);
+  if (priv->head < 0 || priv->head >= priv->max_size)
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA, "Pointer to "
+                   "head of circular file must lie in range [0, %"
+                   G_GUINT64_FORMAT "), but was %" G_GOFFSET_FORMAT ".",
+                   priv->max_size, priv->head);
+      return FALSE;
+    }
+
+  guint64 little_endian_prev_max_size =
+    g_key_file_get_uint64 (priv->metadata_key_file, METADATA_GROUP_NAME,
+                           MAX_SIZE_KEY, &local_error);
+  if (local_error != NULL)
+    goto handle_failed_read;
+
+  guint64 prev_max_size =
+    swap_bytes_64_if_big_endian (little_endian_prev_max_size);
+  return (prev_max_size == priv->max_size) ||
+         resize (self, prev_max_size, error);
+
+handle_failed_read:
+  g_propagate_error (error, local_error);
+  return FALSE;
+}
+
+static void
+emer_circular_file_initable_init (GInitableIface *iface)
+{
+  iface->init = emer_circular_file_may_fail_init;
+}
+
+/* Returns a new circular file or NULL on error. If a circular file does not
+ * already exist at the given path, a new one is created. Limits the physical
+ * size of the underlying data file to max_size bytes. If a circular file with a
+ * different maximum size already exists at the given path, its maximum size is
+ * changed to the given value, which may result in data loss.
+ */
+EmerCircularFile *
+emer_circular_file_new (const gchar *path,
+                        guint64      max_size,
+                        GError     **error)
+{
+  return g_initable_new (EMER_TYPE_CIRCULAR_FILE,
+                         NULL /* GCancellable */,
+                         error,
+                         "path", path,
+                         "max-size", max_size,
+                         NULL);
+}
+
+/* Appends the given element in-memory only. Use emer_circular_file_save to
+ * flush all appended elements. This allows for batching of writes. Note that
+ * elements can not be read with emer_circular_file_read until they have been
+ * saved. Returns TRUE if the given element was successfully appended and will
+ * fit in the space allotted to the circular file. Returns FALSE otherwise.
+ */
+gboolean
+emer_circular_file_append (EmerCircularFile *self,
+                           gconstpointer     elem,
+                           guint64           elem_size)
+{
+  EmerCircularFilePrivate *priv =
+    emer_circular_file_get_instance_private (self);
+
+  guint64 elem_size_on_disk = sizeof (elem_size) + elem_size;
+  guint64 total_size = priv->size + priv->write_buffer->len + elem_size_on_disk;
+  if (total_size > priv->max_size)
```

Comment:
`append` and `save` are separate so that we can avoid writing each variant to disk individually when we flush to the persistent cache every hour. Roddy felt pretty strongly that there should be as little writing to disk as possible in the metrics system so that it doesn't wear out the SD card or eMMC storage. I don't think I understand what you have in mind though.